### ISSUE TITLE
XXE DoS within SVG Parsing

### ISSUE THREAD

**Issue by ProDigySML** [BigBadaboom/androidsvg#122]
# Issue
androidsvg is vulnerable to XXE attacks as some dangerous features are not disabled. This leads to a confirmed denial of service scenario (https://en.wikipedia.org/wiki/Billion_laughs_attack) and may lead to execution of commands on the server.

This issue occurs in the SVG parse section of the code:
https://github.com/BigBadaboom/androidsvg/blob/1f53713100a0e3e30bb3b29308588fc2ea41b06d/androidsvg/src/main/java/com/caverock/androidsvg/SVGParser.java#L576

Prior to parsing the XML, features like entities are not disabled. These should not be required at all within an SVG file.
https://github.com/BigBadaboom/androidsvg/blob/1f53713100a0e3e30bb3b29308588fc2ea41b06d/androidsvg/src/main/java/com/caverock/androidsvg/SVGParser.java#L606

# Remediation
Implementing something down the lines of the following:
```Java
SAXParserFactory spf = SAXParserFactory.newInstance();
spf.setFeature("http://xml.org/sax/features/external-general-entities", false);
spf.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
spf.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
```
This ensures that the entities are no longer parsed and external dtd files are not either.

**Comment by BigBadaboom** [comment]
Thanks for the report.

Entities do have their uses.  The SVG spec itself contains examples that use entity expansion. I am reluctant to completely disable them. Especially considering none of the browsers have either.

But it does make sense to either:
(a) check whether the Android SAX parser has an entity expansion limit. If it does, check that it is set appropriately, or
(b) if we can't set a limit, then disable entity expansion it, but allow the user to enable it if necessary.

**Comment by ProDigySML** [comment]
Fair enough! Thanks for that! Didn't actually know svgs use entities. Both approaches seem like a good idea! Thanks for being so prompt!

**Comment by BigBadaboom** [comment]
Investigations have shown that the expat parser built into Android (which is used by the SAXParser which AndroidSVG uses) does not currently support any limits on entity expansion.  It is still an open bug.

https://github.com/libexpat/libexpat/issues/34

Although there is a way to turn off external entity expansion, there doesn't seem to be a way to turn off the expansion of internal entities. Internal entities are those defined in the XML file itself.  Expat doesn't seem to support the `startEntity()`, `endEntity()`, `internalEntityDecl()`, and the other entity-related SAX methods either.  So there doesn't even seem to be a way to intercept them, before expansion happens.

It might mean that I either:
(a) have to move to the XmlPullParser, which I believe doesn't support entities at all, or
(b) Add a security warning to the documentation, reminding users to not trust external SVG files

**Comment by ProDigySML** [comment]
To be honest, considering there isn't much easily doable, adding a security warning seems reasonable. (Bonus if it's in red :) )

**Comment by BigBadaboom** [comment]
Fixed.  This change will be in the 1.3 release.

I have implemented a dual XML parser approach.  The SVG parser now starts out using the faster XMLPullParser (which doesn't support entities). If it detects entity declarations, it will reset the input stream and switch to the SAX parser (which does support entities).

If you want to disable entity parsing, for security reasons, you can now call the new method `setInternalEntitiesEnabled(false)`

**Comment by TWiStErRob** [comment]
@BigBadaboom Is there a way to always use the one that supports entity parsing? I use it very extensively and don't want to pay the price of resetting the stream.

**Comment by BigBadaboom** [comment]
@TWiStErRob Not at the moment.  If you are finding that this change has had a significant detrimental impact on your app's performance, please open a new issue.

**Comment by TWiStErRob** [comment]
Haven't tested yet, maybe just being paranoid; I'll keep an eye out. Thank you!

**Comment by jaya-shankar** [comment]
@BigBadaboom — we’re currently using version 1.4 of this library in our app. Our recent security report flagged it as potentially vulnerable to an XXE attack (CWE-611).

Could you confirm if this issue has been addressed in the latest version?

**Comment by BigBadaboom** [comment]
@jaya-shankar  This was fixed in release 1.3.  By default XXE processing has been disabled.  So we should not be vulnerable to those sort of attacks.  Unless you deliberately enable it.

Note that CWE-611 predates this fix.  If you have any more details from your security report on why this may be still vulnerable, please let me know.
